### ISSUE TITLE
setting TOLUD on X230 (or maybe a related option)

### ISSUE THREAD

**Issue by Abraxas-Knister** [n4ru/1vyrain#53]
I want to change the TOLUD to a lower setting but cant find the setting in the "Advanced" menu. I'd alternatively want to activate an option related to "Above 4G decoding", but can't find that either. Here's why I want that (I don't know if any of the following matters but I include it anyways):

I want to set up an eGPU with an EXP GDC (not sure which) and a NVIDIA GTX 750 (not ti). I'm on linux lubuntu 18.04, kernel is 4.15.0-117-generic, boot is GPT UEFI (grub2), BIOS is 2.77, flashed via 1vyrain at Sep 4 (a week ago). When I plug in the expresscard attached to the EXP GDC I can see these things (https://pastebin.com/fibRDz9Z) in the kernel log (dmesg), it should be exactly everthing that happens when I plug in the card. When I install the nvidia driver I see this (https://pastebin.com/KF1nUXzh) repeated over and over again (I chose a random start, but it seems sensible to me--since it is annoying to watch at and the driver doesn't do anything if the GPU isn't recognised the driver is currently not installed).

Upon researching on this I was told to

- enable "above 4G decoding" or

- smaller the TOLUD

- do a BIOS upgrade

- booth with UEFI

so I searched for the options and did the other two things. I upgraded from the 1.X BIOS I still had at this point to 2.77, where I then found out that this doesn't change anything and that changing the TOLUD wasn't an option, so I asked further and was directed to 1vyrain which also couldn't resolve it.

_Note: to install I downgraded from stock 2.77 to 2.6 and booted the 1vyrain image via UEFI, I forgot to do a BIOS reset though, and I did not follow through all of (https://github.com/gch1p/thinkpad-bios-software-flashing-guide), just the downgrading part. Actually I thought of this the second I was prompted to make a decision and pressed 3 for a reboot a minute or five after but it installed anyways (why tho? was It because I waited too long? It said something about a timeout but I don't remember entirely). I don't know how to check if the install was fully succesful but I don't see anything odd (if not, can I safely just do it again, and do I need to do the downgrade again?)_

Now I'm asking here since nobody seems to know an answer (and I'm sufficiently desperately undereducated in CS stuff around BIOS, kernel and hardware that I don't even know where to begin to read myself to a solution). At this point I wanted to add that it probably needs to be verified that the problem is with the BIOS and I'm not sure if this then is an issue of 1vyrain or rather a feature request. I'm also not sure how much of this is misplaced, I don't want to ask for more here than confirmation that it is a BIOS problem and a way to resolve it, if it's a kernel problem or a problem nouveau and nvidia somehow share, I'll of course accept that and try my luck with them (as frustrating it may be to not have a solution).

To resolve the issue I was told a few other things, adding these too, to make is easier to troubleshoot if this really is an issue with the BIOS:
 
- add kernel parameter "pci=nocrs", this doesn't change anything maybe other kernel parameters might help?

- different (newer) kernel: the sources for bionic beaver have 5.0, 5.3 and 5.4, I tried out the latest 5.3 they had but it didn't work on that either.

- DSDT override (didn't, frightens me)

- peculiarly other X230 users had working eGPUs (and one of these had a GTX 750 ti) without needing to alter the BIOS at all

**Comment by digmorepaka** [comment]
The eGPU usually just works on these machines, have you tried clearing setup settings to default?

**Comment by Abraxas-Knister** [comment]
Didn't change anything of the config the 1vyrain installed with. I presume it copies the config that is already there, in that case I'm going to try a reset asap. Is there a possibility that me not doing a reset before the install could have led to the install not being complete? Ie that the "Advanced" menu isn't unlocked for everything because some exploits shouldn't be made? 

I read up on eGPU with the X230 before I bought my stuff and it sounded all quite positive without even mentioning that UEFI would even be necessary, so I'm a bit puzzled myself =]

**Comment by digmorepaka** [comment]
1vyrain flashes a clean binary. All settings are reset, however I am not sure if they are the "Optimized defaults" or not. You should reset them.

**Comment by Abraxas-Knister** [comment]
It said they aren't activated. I've resetted it and then changed the boot preference from "both" and "Legacy first" to "UEFI only" (since my system apparently boots both via UEFI and legacy, I don't know what is going on here). I did a startup via legacy anyways.

When I'm doing the boot with the device attached, it wouldn't even list it in `lspci`, I have to pull out the expresscard and attach it again. Same kernel log however, even after the reset.

**Comment by Abraxas-Knister** [comment]
Can I read out what the TOLUD is set to even if the BIOS doesn't have an option for it? If so can I also write it? I've been told repeatedly that this setting has the wrong value if the card doesn't work and everything else is OK, so I would like to verify the value of the setting.

**Comment by digmorepaka** [comment]
Some forums suggest booting up with the eGPU connected, disabling it in device manager, unplugging it, and plugging it back in. Have you tried this sequence?

Investigating the TOLUD setting at the moment.

**Comment by digmorepaka** [comment]
@Abraxas-Knister the "above 4g decoding" is in memory configuration as "memory remap" https://lewd.pics/p/WCs2.jpg (sorry for terrible image quality... 45$ phones don't tend to have good cameras :P)

**Comment by Abraxas-Knister** [comment]
The word alone is good enough to find it. I actually did find it before and wodered if it was connected. I don't remember trying it out.

I dont know what the device manager would be, I don't (know if I) have a graphical program for that. The eGPU is not found if I boot while it is connected, `lspci` doesn't list it, that is. I think I would then not be able to disable it in the device manager.

**Comment by Abraxas-Knister** [comment]
Looked at the BIOS config and the "memory remap" is enabled by default. In "system agent (SA) configuration" I found a "graphics configuration" with values I will make a photo of.

**Comment by Abraxas-Knister** [comment]
![gc](https://user-images.githubusercontent.com/43271791/93149377-70053a80-f6f6-11ea-85c2-b300bea66598.jpg)
![igd](https://user-images.githubusercontent.com/43271791/93149378-709dd100-f6f6-11ea-8fa4-c43dfc0f2ae0.jpg)

**Comment by Abraxas-Knister** [comment]
Most of these settings are only relevant for internal graphics. I looked around and it seems the "GTT size" should be much larger than 2MB but this is already the max.

I'm still unsure whether this is all really a BIOS problem. I mean I don't understand the settings since I don't know detailed what they control, but you said it should work if the BIOS is set to default values. So I guess the problem is either with the hardware itself (doesn't seem like it, the fan works and I can get hardware info from `lspci`) or the kernel (modules).

**Comment by digmorepaka** [comment]
Have you tried another maybe simpler pcie device? For example a network card or a maybe even a different gpu.

**Comment by Abraxas-Knister** [comment]
I don't have one at hand. I would like to test the onboard audio device, since that seems to be installed correctly if I'm reading that kernel log correctly. I don't know how to, though. I think the next try would be to do an install of a different distribution (arch) or to try a mainline kernel. I don't have high hopes a computer shop could help me out here, I'd think 1vyrain is as specialised on X230 as it gets. Of course installing windows would be an option too, but not a favoured one =]

**Comment by Abraxas-Knister** [comment]
I've been to a shop and came back with two ideas (and the remark that they would certainly not immediately know what to do and thus just cost me a bit).

- try out it the problem comes up while trying the setup with a life distribution. I tried lubuntu 20.04 and Pop!_OS 20.04, both didn't work (I picked the NVIDIA alternative of Pop!_OS)

- probably it's the EXP GDC

the first hints that the problem is most probably with either the hardware or the BIOS and not with the kernel or driver, the second seems likely. Pop!_OS failed in an interesting way, the nvidia driver didn't cycle through the messages over and over again, it only showed them once.

I've found that the adapter has two physical switches, one for "PTD" with the values "15S", "7S" and "OFF", it was delivered with  "OFF" and some guide suggested "7S", the other switch is for "ATX" with the values "OFF" and "ON", it is sealed and stands at "OFF". Neither tells me anything but I tried all three "PTD" options without touching the "ATX" options (without luck).

I'll probably direct my woes to people knowledgable about the EXP GDC BEAST V8.5c (just found out which version I have) as I 1vyrain most probably is not the thing that is in the wrong here.
